### ISSUE TITLE
[Bug]: Soft hyphens showing up in headers

### ISSUE THREAD

**Issue by AJDol** [Sigil-Ebook/Sigil#748]
### Bug Description

![Screenshot_20240403_145956](https://github.com/Sigil-Ebook/Sigil/assets/166004752/279a370f-139f-4830-8217-9ab639cec682), Soft hyphens in headers of my cookbook are showing in Sigil Preview window and Kobo e-readers. This happens only in headers, not in other text on the page.
The used font is: [Almaq Rough](https://www.maisfontes.com/almaq-rough.font).
Soft hyphens are inserted with the Calibre Polishing function in the entire text, not only the headers.

### Platform (OS)

Windows (Default)

### OS Version / Specifics

Windows 10 Pro 22H2 19045.4170

### What version of Sigil are you using?

Since 2.1.0, 2024.04.01 15:40:18 UTC on Qt 6.6.2

### Any backtraces or crash reports

```shell
none
```

**Comment by AJDol** [comment]
Hallo Becky ,

 

I respectfully disagree. Why are the soft-hyphens in the last word not showing?

The behavior in the Sigil viewer is erratic, sometimes showing hyphens, sometimes not.

When I remove one soft-hyphen, another is showing up in the same word.

Next time a reload a page a new soft-hyphen is visible.

There must by an issue in the new version of Sigil, because I’ve never noticed this before with the same font.

 

Best regards,

Antoni

**Comment by BeckyDTP** [comment]
> Why are the soft-hyphens in the last word not showing?
I might suspect that this is the whole idea of the problem. A certain randomness in the display of just that particular font.
When I remove the first shy-hyphen, I get this view:
![image](https://github.com/Sigil-Ebook/Sigil/assets/22416021/0803be15-1d2a-42ac-bff8-3ddb10013d8a)

or this:
![image](https://github.com/Sigil-Ebook/Sigil/assets/22416021/fd2179ae-5d69-49fe-ab1f-6dc2073610bb)



I work in Sigil every day, and if such a problem occurred in other fonts, I would certainly notice it.
But of course it could be something unique to Qt6.

**Comment by dougmassay** [comment]
You also mentioned Kobo was doing the same thing. It's unlikely that Sigil AND Kobo broke at the same time.

It's simple enough to test.

Try the same code with no embedded font and try the same code with a different embedded font (one with Unicode support for the soft hyphen character).

**Comment by AJDol** [comment]
Hello Backy,

 

These double hyphens are interesting. I don’t think Calibre would enter double hyphens in a header. May Sigil now does under certain conditions?

For me, this is also the first time I encounter this problem, and I’d never seen it before I installed this last version.

So, I cannot be the font. I must be Sigil. A change in the hyphenation routine maybe? A new dictionary?

I would be happy to adjust my process if I know wat Sigil is doing.

Please escalate this issue to someone who knows what is changed in Sigil 2.1.0 or perhaps Qt 6.6.2? My prior install was Sigil-2.0.2.

 

Thanks in advance,

Antoni

**Comment by BeckyDTP** [comment]
It could be something with Qt6 though, because I checked on 1.9.1 (Qt 5.12.9) and everything is OK there.

![image](https://github.com/Sigil-Ebook/Sigil/assets/22416021/3e5bccd9-aee2-476e-b922-a89a4b644f63)

**Comment by BeckyDTP** [comment]
2.0.2 (issue exist):
![image](https://github.com/Sigil-Ebook/Sigil/assets/22416021/6fbd806e-8716-4f09-9eb5-46ba4cb88cd8)

**Comment by dougmassay** [comment]
I'm still not getting the Kobo ereader connection.

Is the font being embedded, or just used in Sigil's Preview preferences?

**Comment by BeckyDTP** [comment]
![Screenshot_2024-04-04-14-26-37-869_com kobobooks android](https://github.com/Sigil-Ebook/Sigil/assets/22416021/ecacdc05-23c3-4c5e-b788-1107f060b9f5)

Screenshot from kobo app on Android. Issue exist.

**Comment by kevinhendricks** [comment]
Is there a test case epub attachment here with that font included.  The original font link is loaded with malware and crap links.

**Comment by dougmassay** [comment]
Also note that the version of Almaq Rough on that website is an evaluation copy only. It's not freely redistributable in commercial endeavors without paying for a license.

**Comment by dougmassay** [comment]
Clearly there's something wrong with the font in my opinion. It's possible that whatever is wrong was not exposed in earlier versions of QtWebEngine in Qt5, but IS exposed in Qt6, but ...

We have no idea how long it's been a problem in Kobo's rendering engine. Or other rendering engines for that matter. The Kobo test is proof that its not just a Sigil, or Qt6 glitch.

**Comment by kevinhendricks** [comment]
If just one of the glyphs for that font for the soft hyphen is filled in, then if that size is closest to the header, the synthesized header font will be incorrect?  Try bolding and changing font sizes, and you should see if that is the case.

**Comment by kevinhendricks** [comment]
This may in fact be a chrome bug.  Perhaps kobo on Android uses chrome?

**Comment by dougmassay** [comment]
I believe Kobo has two different rendering engines. One for epubs and one
for their proprietary kepub format. One or both could use chrome.

**Comment by kevinhendricks** [comment]
Either way, I grabbed that font from that crappy download site, then loaded it into FontForge and it shows a "hyphen" glyph for the unicode character 173, 0xAD (soft hyphen) in the cmap.  That character should *not* in fact have any glyphs as it is a zero width placeholder for possible hyphenation.  Some fonts do include a glyph there but then mark the character as unprintable.  As far as I can tell, it is not marked as such.

So whatever the problem, it is related to this font and how it is processed by various font engines.

**Comment by AJDol** [comment]
Thx for your research. I will have to account for that. For now, all soft hyphens are removed from headers.

**Comment by kevinhendricks** [comment]
To test things, I used Fontforge to clear the glyph for the soft hyphen and then save it to the font.
I then ran the font on a test case with soft hyphens and it now works and hyphenates correctly in Sigil / Chrome.

So I do believe that some font rendering systems, ignore glyphs when they exist in the soft hyphen location while other font renderers use what is there regardless.

I have no idea which is technically more correct.

BTW, FontForge found lots of other errors and warnings when operating with this font such as improperly mapped glyphs/cmaps, unclosed points, etc.  I recommend staying far away from this font or at least fix all those errors and warnigns using a font enditing and checking tool like FontForge.  What editing it would do to the license for this font is then unknown.

So I am closing this since it appears to be issues with that Font that cause some font renderers problems.
